java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -767,4 +799,52 @@ public String getClientLibToken() {
   public String getQuotaProjectId() {
     return quotaProjectId;
   }
+
+  /**
+   * Returns the resolved host for the Service to connect to Google Cloud
+   *
+   * <p>The resolved host will be in `https://www.{serviceName}.{resolvedUniverseDomain}/` format.
+   * The resolvedUniverseDomain will be set to `googleapis.com` if universeDomain is null.
+   */
+  @InternalApi
+  public String getResolvedHost(String serviceName) {
+    if (universeDomain != null && universeDomain.isEmpty()) {
+      throw new IllegalArgumentException("Universe Domain cannot be empty");
+    }
+    String resolvedUniverseDomain =
+        universeDomain != null ? universeDomain : Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    // The host value set to DEFAULT_HOST if the user didn't configure a host. If the
+    // user set a host the library uses that value, otherwise, construct the host for the user.
+    // The DEFAULT_HOST value is not a valid host for handwritten libraries and should be
+    // overriden to include the serviceName.
+    if (!host.equals(DEFAULT_HOST)) {
```

Comment:
This could throw null pointer exception if host is null. I see that host is set to DEFAULT_HOST in the constructor, but I would change the code to the following to prevent any possible NullPointerException in the future.
```suggestion
    if (!DEFAULT_HOST.equals(host)) {
```